stdlib/universe/kaufmansAMA.go

```diff
@@ -0,0 +1,259 @@
+	ns := new(KamaProcedureSpec)
+	*ns = *s
+	return ns
+}
+
+// TriggerSpec implements plan.TriggerAwareProcedureSpec
+func (s *KamaProcedureSpec) TriggerSpec() plan.TriggerSpec {
+	return plan.NarrowTransformationTriggerSpec{}
+}
+
+func createkamaTransformation(id execute.DatasetID, mode execute.AccumulationMode, spec plan.ProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	s, ok := spec.(*KamaProcedureSpec)
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid spec type %T", spec)
+	}
+	cache := execute.NewTableBuilderCache(a.Allocator())
+	d := execute.NewDataset(id, mode, cache)
+	t := NewkamaTransformation(d, cache, s)
+	return t, d, nil
+}
+
+type kamaTransformation struct {
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+
+	n       int64
+	columns []string
+}
+
+func NewkamaTransformation(d execute.Dataset, cache execute.TableBuilderCache, spec *KamaProcedureSpec) *kamaTransformation {
+	return &kamaTransformation{
+		d:     d,
+		cache: cache,
+
+		n:       spec.N,
+		columns: spec.Columns,
+	}
+}
+
+func (t *kamaTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func (t *kamaTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	builder, created := t.cache.TableBuilder(tbl.Key())
+	if !created {
+		return fmt.Errorf("KAMA found duplicate table with key: %v", tbl.Key())
+	}
+	cols := tbl.Cols()
+	doKAMA := make([]bool, len(cols))
+	for j, c := range cols {
+		found := false
+		for _, label := range t.columns {
+			if c.Label == label {
+				if c.Type != flux.TInt && c.Type != flux.TUInt && c.Type != flux.TFloat {
+					return fmt.Errorf("cannot take KAMA of column %s (type %s)", c.Label, c.Type.String())
+				}
+				found = true
+				break
+			}
+		}
+
+		if found {
+			mac := c
+			mac.Type = flux.TFloat
+			_, err := builder.AddCol(mac)
+			if err != nil {
+				return err
+			}
+			doKAMA[j] = true
+		} else {
+			_, err := builder.AddCol(c)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return tbl.Do(func(cr flux.ColReader) error {
+		if cr.Len() == 0 || len(cr.Cols()) == 0 {
+			return nil
+		}
+
+		for j, c := range cr.Cols() {
+			if !doKAMA[j] {
+				for i := int(t.n); i < cr.Len(); i++ {
+					if err := builder.AppendValue(j, execute.ValueForRow(cr, i, j)); err != nil {
+						return err
+					}
+				}
+			} else {
+				var values []float64
+				switch c.Type {
+				case flux.TInt:
+					arr := cr.Ints(j)
+					for i := 0; i < arr.Len(); i++ {
+						values = append(values, float64(arr.Value(i)))
+					}
+				case flux.TUInt:
+					arr := cr.UInts(j)
+					for i := 0; i < arr.Len(); i++ {
+						values = append(values, float64(arr.Value(i)))
+					}
+				case flux.TFloat:
+					arr := cr.Floats(j)
+					values = arr.Float64Values()
+				}
+				kers := t.makeKERS(values, int(t.n))
+				var kama float64
+				sc := 0.0
+				kama = values[t.n-1]
```

Comment:
if `t.n` is greater than the number of rows in the table, then this will `panic`, right?  What would the best behavior for this be?  It could either return an error, or just return an empty table.  What does EMA do?

Reply:
EMA doesn't consider this case.